### ISSUE TITLE
Expected AWS S3 Credential Locations are Inconsistent Across `nix` Commands for Daemon Users

### ISSUE THREAD

**Issue by nrdxp** [NixOS/nix#5723]
**Describe the bug**

I have found some issues with how Nix reads AWS Credentials while setting up a `post-build-hook` that uploads to an S3 cache in a CI action. This particular action uses the same cache to pull already built resources, and the cache is private and so requires credentials which is how I found the problem:

* For commands that interact with the `nix-daemon`, such as `nix build`, the aws credentials file must exist at `/root/.aws/credentials`. Since there is no way to modify the daemon's environment after it starts, there is no way to pass in any `AWS_*` vars that would usually change this bahavior.

* For commands that don't interact with the daemon, such as `nix copy` the credentials are able to be read from the standard AWS credentials chain provided by the AWS SDK. Importantly, these come from the _callers_ environment. So one can set `AWS_ACCESS_KEY_ID` and `AWS_SECRET_ACCESS_KEY` in the environment, call` nix copy --to s3://<some-cache>` and have it work out.

**Steps To Reproduce**

1. setup a private s3 bucket that requires credentials to push _and_ pull
2. setup your AWS credentials for your calling user
3. run `nix copy --to s3://<your-cache>`, if credentials are set properly it should work
4. Setup this same cache as a substituter in your `nix.conf`
5. run `nix build` on the same path you just uploaded (you may have to manually delete it to rebuild it). You should get a AWS warning, and the substitution will not take place.
6. now setup a proper credentials file at `/root/.aws/credentials` and try again, this time it should work

**Expected behavior**

We should only have to set AWS vars in one place for all commands. Since `nix copy` runs as calling user, while `nix build` runs as daemon user, this is currently impossible without setting credentials in at least two places. Perhaps a new `nix.conf` setting is needed to point to a single credentials file? Although personally, I would prefer it if I could dynamically pass in vars like `AWS_ACCESS_KEY_ID` to `nix build`, but that may be much trickier.

**`nix-env --version` output**

`nix (Nix) 2.5pre20211203_2e606e8`

**Additional context**

If you don't know how to setup an S3 cache to test this behavior, feel free to review the [docs on it](https://github.com/NixOS/nix/blob/master/doc/manual/src/package-management/s3-substituter.md).

**Comment by nrdxp** [comment]
I'm going to close for now, as I seem to have mistaken something here. Some of my assumptions here appear to be incorrect and reproduction doesn't work as outline. I've got to do further testing to provide a more useful analysis.

**Comment by pinpox** [comment]
> * Importantly, these come from the _callers_ environment. So one can set `AWS_ACCESS_KEY_ID` and `AWS_SECRET_ACCESS_KEY` in the environment, call` nix copy --to s3://<some-cache>` and have it work out.

This doesn't work for me. I've set `AWS_SECRET_ACCESS_KEY` and `AWS_ACCESS_KEY_ID` but `nix copy` ignores those variables. Configuring them in `~/.aws/credentials` works, but is not really an option when trying to copy from a CI job

**Comment by nrdxp** [comment]
I'm reopening for now as this is still an issue. Not sure exactly which commands talk to daemon and which don't but I have shown that `nix store ping` does work with only local user credentials but `nix build` does not unless a credentials file exists for the root user, so that's at least a starting point for reproducing. Perhaps the `nix.conf` needs a setting to point to a credential file location?

**Comment by kgtkr** [comment]
I solved it by doing `sudo cp -R .aws /var/root/.aws`.
But I want a simple solution

**Comment by nrdxp** [comment]
> I solved it by doing `sudo cp -R .aws /var/root/.aws`.

Yes, copying credentials to the root user does work, but it's more of a work around than a real solution. I was thinking maybe we need a new nix.conf setting for AWS creds. One can also set `AWS_*` env vars in the systemd service running NixOS as another (potentially cleaner) workaround.

**Comment by kgtkr** [comment]
If there is an item in nix.conf that specifies aws credentials like `netrc-file`, will it be resolved?
aws authentication is not compatible with netrc. so netrc is not available.
I worked around it by disabling minio authentication, deploying a basic auth proxy and configuring netrc, but it's difficult to use this method in s3.

**Comment by nrdxp** [comment]
> If there is an item in nix.conf that specifies aws credentials like `netrc-file`, will it be resolved?

No, I'm saying one solution to the problem might be to add one.

**Comment by codygman** [comment]
This bug is making trying out nix for my company very difficult since we use a private s3 as our binary cache and have multi-factor auth. Sadly *it's currently not feasible to use a public cache*.

There might be ways around it, but that means I have to manually generate a session token every time it expires and update `/var/root/.aws/credentials`.

I don't think nix will get very much adoption in my company if that's a required step to get cached builds.

**Comment by codygman** [comment]
Oh... quick idea:

The [docs](https://github.com/NixOS/nix/blob/master/doc/manual/src/package-management/s3-substituter.md) should probably mention this issue as well as give a synopsis of the `nix build` uses daemon (root), nix copy uses client, and that it can lead to confusing issues like this.

**Comment by blaggacao** [comment]
@codygman see: https://github.com/cachix/install-nix-action/pull/150

**Comment by endgame** [comment]
@codygman AIUI the official AWS solution for this problem is [IAM Roles Anywhere](https://aws.amazon.com/about-aws/whats-new/2022/07/aws-identity-access-management-iam-roles-anywhere-workloads-outside-aws/), but if you don't have PKI set up then the AWS-flavoured version (Certificate Manager) looks like it gets expensive.

Could you stand up a small EC2 instance or something that has access to a role (so you can use an instance profile instead of long-lived keys and SSO) and pushes the AK/SK/Token to your Mac whenever things change?
